File: 2021_13.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include "grid.h"


#ifdef DEBUG
#define DBGPRINT(...) fprintf(stderr, __VA_ARGS__)
#else
#define DBGPRINT(...)
#endif

#define AOC_DAY 13
#define AOC_YEAR 2021


#ifndef TEST
#define GRID_WIDTH 1400
#define GRID_HEIGHT 1400
#else  /* TEST */
#define GRID_WIDTH 11
#define GRID_HEIGHT 15
#endif

void read_points(struct grid * g)
{
    int scanf_retval;
    int x, y;

    scanf_retval = scanf("%d,%d", &x, &y);
    while(2 == scanf_retval)
    {
        if(x >= g->width || y >= g->height)
        {
            printf("ERROR: found point (%d, %d) off grid of size (%d, %d)\n",
                    x, y, g->width, g->height);
            exit(1);
        }
        /* mark this point on the grid */
        grid_set_value(g, x, y, 1);
        scanf_retval = scanf("%d,%d", &x, &y);
    }
}

int read_folds(struct grid * g, unsigned int max_folds)
{
    char axis;
    int line;
    int scanf_retval;
    int foldn;
    int x, y;

    for(foldn=0; foldn < max_folds; foldn++)
    {
        scanf_retval = scanf("fold along %c=%d\n", &axis, &line);
        if(2 != scanf_retval)
        {
            foldn--;
            break;
        }
        if('x' == axis)
        {
            /* for each row */
            for(y=0; y < g->height; y++)
            {
                /* for each point on this row after the fold line */
                for(x = 0; x < g->width - line ; x++)
                {
                    if(1 == grid_get_value(g, line + x, y))
                    {
                        /* if this is a point, fold it */
                        grid_set_value(g, line - x, y, 1);
                        grid_set_value(g, line + x, y, 0);
                    }
                }
            }
        }
        else if('y' == axis)
        {
            /* for each column */
            for(x=0; x < g->height; x++)
            {
                /* for each point on this column below the fold line */
                for(y = 0; y < g->width - line ; y++)
                {
                    if(1 == grid_get_value(g, x, line + y))
                    {
                        /* if this is a point, fold it */
                        grid_set_value(g, x, line - y, 1);
                        grid_set_value(g, x, line + y, 0);
                    }
                }
            }
        }
    }
    return foldn;
}

int main(int argc, char ** argv)
{
    struct grid g;

    create_grid(&g, GRID_WIDTH, GRID_HEIGHT, 0);
#ifdef DEBUG
    print_grid(&g);
#endif

    /* fill in initial points */
    read_points(&g);
#ifdef DEBUG
    print_grid(&g);
#endif

    /* read fold marks */
    read_folds(&g, 1);
    /* count the number of dots */
    printf("After first fold, found %lu dots\n", sum_grid(&g));

    printf("made %u folds\n", read_folds(&g, (unsigned int) -1));
    print_grid_section(&g, 0, 40, 0, 6);

    for(int y = 0; y < 6; y++)
    {
        for(int x = 0; x < 40; x++)
        {
            if(1 == grid_get_value(&g, x, y))
                printf("#");
            else
                printf(" ");
        }
        printf("\n");
    }

    free_grid(&g);

    return 0;
}

File: 2021_13.rb
require 'logger'
require 'scanf'
require_relative 'lib/grid'

AOCDAY = 13
AOCYEAR = 2021

# count number of dots
def sumfunc(sum, n)
  sum += n == '#' ? 1 : 0
end

def run(inputfile, logger, **kwargs)
  dots = []
  folds = []

  File.open(inputfile) do |fh|
    # read dot coordinates
    line = fh.readline
    while line.strip.length > 0
      dots.append(line.strip.split(',').map(&:to_i))
      line = fh.readline
    end

    # read folds
    fh.each_line.each do |line|
      folds.append(line.strip.scanf('fold along %c=%d'))
    end
  end

  # create grid and apply dots
  xmax = dots.max { |point0, point1| point0[0] <=> point1[0] } [0] + 1
  ymax = dots.max { |point0, point1| point0[1] <=> point1[1] } [1] + 1
  grid = Grid.new(width: xmax, height: ymax, default: ' ', delimiter: '')

  dots.each do |x,y|
    grid.set(x, y, '#')
  end

  # fold   result: scalar - (y - scalar) = 2 * scalar - y
  folds.each_with_index do |axisnscalar, nfolds|
    axis, scalar = axisnscalar
    logger.debug("Folding at #{axis}=#{scalar}");
    if 'x' == axis
      # hopefully no folds where scalar < ymax/2
      (scalar...xmax).each do |x|
        (0...ymax).each do |y|
          if '#' == grid.get(x,y)
            logger.debug("Folding (#{x},#{y}) to (#{2 * scalar - x},#{y})");
            grid.set(2 * scalar - x, y, '#')
            grid.set(x, y, ' ');
          end
        end
      end
      xmax = scalar + 1
    else     # y axis
      (scalar...ymax).each do |y|
        (0...xmax).each do |x|
          if '#' == grid.get(x,y)
            logger.debug("Folding (#{x},#{y}) to (#{x},#{2 * scalar - y})");
            grid.set(x, 2 * scalar - y, '#')
            grid.set(x, y, ' ');
          end
        end
      end
      ymax = scalar + 1
    end
    logger.debug("\n" + grid.to_s_subgrid(xmax:xmax, ymax:ymax) + "\n");
    if 0 == nfolds
      puts("#{grid.sum_subgrid(xmax:xmax, ymax:ymax, p: :sumfunc)} dots after 1st fold")
    end
  end

  sum = grid.sum_subgrid(xmax:xmax, ymax:ymax, p: :sumfunc)
  puts("Found #{sum} marked spots");

  # print
  puts grid.to_s_subgrid(xmax:xmax, ymax:ymax);
end


if __FILE__ == $0
  require 'optparse'

  options = {}
  OptionParser.new do |opts|
    opts.banner = "Usage: $0 [options] <inputfile> [<more inputfiles>]\n" \
      "  Advent of Code #{AOCYEAR} day #{AOCDAY}"

    opts.on("-v", "--verbose", "Be verbose") do |v|
      options[:verbose] = v
    end

    opts.on("-2", "--part2", "Run part 2") do |part2|
      options[:part2] = true
    end

    opts.on("-h", "--help", "Print this help message") do
      puts opts
      exit
    end
  end.parse!

  logger = Logger.new(STDOUT);
  logger.level = (options.include? :verbose) ? Logger::DEBUG : Logger::WARN

  ARGV.each do |inputfile|
    run(inputfile, logger, **options)
  end
end
